File: go/earthbucks_lib/pkg/tx_signature.go
package earthbucks

import "errors"

// Constants for hash types and size
const (
	SIGHASH_ALL          = 0x00000001
	SIGHASH_NONE         = 0x00000002
	SIGHASH_SINGLE       = 0x00000003
	SIGHASH_ANYONECANPAY = 0x00000080
	TxSignatureSize      = 65
)

// TxSignature represents a transaction signature
type TxSignature struct {
	hashType *U8
	sigBuf   *FixedBuf
}

// NewTxSignature creates a new TxSignature
func NewTxSignature(hashType *U8, sigBuf *FixedBuf) *TxSignature {
	return &TxSignature{hashType: hashType, sigBuf: sigBuf}
}

// ToBuf serializes the TxSignature into a byte buffer
func (tx *TxSignature) ToBuf() []byte {
	buf := make([]byte, TxSignatureSize)
	buf[0] = tx.hashType.value
	copy(buf[1:], tx.sigBuf.buf)
	return buf
}

// FromBuf deserializes a byte buffer into a TxSignature
func TxSignatureFromBuf(buf []byte) (*TxSignature, error) {
	if len(buf) != TxSignatureSize {
		return nil, errors.New("invalid TxSignature length")
	}
	hashType, err := NewU8(buf[0])
    if  err != nil {
        return nil, err
    }
	sigBuf, err := NewFixedBuf(64, buf[1:])
	if err != nil {
		return nil, err
	}
	return NewTxSignature(hashType, sigBuf), nil
}

File: ts/earthbucks-lib/src/tx-signature.ts
import { FixedBuf, WebBuf } from "./buf.js";
import { U8, U16, U32, U64 } from "./numbers.js";

export class TxSignature {
  static readonly SIGHASH_ALL = new U8(0x00000001);
  static readonly SIGHASH_NONE = new U8(0x00000002);
  static readonly SIGHASH_SINGLE = new U8(0x00000003);
  static readonly SIGHASH_ANYONECANPAY = new U8(0x00000080);
  static readonly SIZE = 65;

  hashType: U8;
  sigBuf: FixedBuf<64>;

  constructor(hashType: U8, sigBuf: FixedBuf<64>) {
    this.hashType = hashType;
    this.sigBuf = sigBuf;
  }

  toBuf(): WebBuf {
    const hashTypeBuf = WebBuf.alloc(1);
    hashTypeBuf.writeUInt8(this.hashType.n);
    return WebBuf.concat([hashTypeBuf, this.sigBuf.buf]);
  }

  static fromBuf(buf: WebBuf): TxSignature {
    if (buf.length !== TxSignature.SIZE) {
      throw new Error("Invalid TxSignature length");
    }
    const hashType = new U8(buf[0] as number);
    const sigBuf = buf.subarray(1);
    const sigFixedEbxBuf = FixedBuf.fromBuf(64, sigBuf);
    return new TxSignature(hashType, sigFixedEbxBuf);
  }
}
